lib/Reconciler.lua

```diff
@@ -467,6 +483,52 @@ local function set(rbx, key, value)
 	rbx[key] = value
 end
 
+--[[
+	Computes a full name, even if the element is not parented to the DataModel yet.
+]]
+local function computeFullName(handle, rbx)
+	-- If the instance is already parented, just use GetFullName.
+	-- Also use GetFullName if the handle is nil.
+	if rbx.Parent or handle == nil then
+		return rbx:GetFullName()
+	-- If the instance doesn't have a parent, it's still being reified, so we
+	-- need to build the full name manually.
+	else
+		local fullName = rbx.Name
+		local level = handle._parentHandle
+
+		-- For the case of root elements, they have no _parentHandle, but they
+		-- do have a _parent, which is the parent that the root is being reified to.
+		if not level then
+			-- We can reify components to nil.
+			if handle._parent then
+				fullName = handle._parent:GetFullName() .. "." .. fullName
+			end
+		end
+
+		while level do
+			local rbx = level._rbx
+			fullName = rbx.Name .. "." .. fullName
+
+			local nextLevel = level._parentHandle
+			-- If there is a _parentHandle element, travel through it.
+			if nextLevel then
+				level = nextLevel
+			-- Otherwise, we've reached the root element of this element tree.
+			-- Use the _parent value and stop the loop.
+			else
+				if level._parent then
+					fullName = level._parent:GetFullName() .. "." .. fullName
+				end
+
+				break
+			end
+		end
+
+		return fullName
+	end
+end
```

Comment:
My gut says that this function can probably be much simpler.

If we add `_parentHandle` to every node (which makes sense to me), then we can have something like:

```lua
local function computeFullName(handle)
	assert(handle._rbx, "Can only computeFullName of a primitive instance handle!")

	local fullName = handle._key

	while handle._parentHandle do
		handle = handle._parentHandle

		if handle._rbx then
			fullName = handle._key .. "." .. fullName
		end
	end

	if handle._parent then
		fullName = handle._parent:GetFullName() .. fullName
	end

	return fullName
end
```